frontend/src/services/authentication_handler_service/authentication_handler_service.ts

```diff
@@ -12,34 +11,67 @@ export class FailureToSignInError extends Error {
 }
 
 /**
- * Communicates with gapi to retrieve auth code from user.
- * Exchanges auth code and redirect URI for ID token from backend.
+ * Communicates with gapi to manage a user's login status
+ * and retrieve auth codes / tokens.
  */
 export class AuthenticationHandlerService {
+  token: string | undefined = undefined;
+  authInstance: gapi.auth2.GoogleAuth | undefined = undefined;
+
   /** Backend is responsible for retrieving the ID token. */
-  constructor(private readonly backend: BackendAuthenticationInterface) {
-    // Safe because the gapi script is loaded in the head with no async/defer.
-    // TODO: Make gapi script async and call this when it has loaded
-    window.gapi.load('auth2', () => {
-      window.gapi.auth2.init({
-        client_id: CLIENT_ID,
-      });
-    });
+  constructor(private readonly backend: BackendAuthenticationInterface) {}
+
+  /**
+   * Determines if a user is logged in.
+   * @return true if user is logged in, false otherwise
+   */
+  getUserLoginStatus(): boolean {
+    const auth = this.getAuthInstance();
+    return auth.isSignedIn.get() && this.getToken() !== undefined;
   }
 
   /**
-   * Signs the user in, retrieving a token from the backend.
+   * Gets the parsed ID token of the user.
+   * @return the parsed token if it exists, otherwise the empty object
+   */
+  getParsedToken() {
+    const token = this.getToken();
+    return token
+        ? JSON.parse(atob(token.split(".")[1]))
+        : {};
+  }
+
+  /**
+   * Gets the ID token of the user.
+   * @return the token if it exists, otherwise undefined
+   */
+  getToken(): string | undefined {
+    const auth = this.getAuthInstance();
+    const response = auth.currentUser.get().getAuthResponse();
+    if (!response.id_token) {
+      return;
+    }
+    return response.id_token;
+  }
+
+  /**
+   * Signs the user in, calls a function on the token retrieved from backend.
    * @param scopes the scopes our app requests for oauth
-   * @return the string of the JWT token retrieved by the backend
+   * @param tokenConsumer a function that takes the token as input and is
+   *     called when the token is received from the backend
    * @throws FailureToSignInError if sign in fails
    */
-  async signIn(scopes: string): Promise<string> {
+  async signIn(scopes: string,
+               tokenConsumer: (token: string) => void): Promise<void> {
```

Comment:
Having both a callback and a promise is definitely an anti-pattern. You can build a promise out of a callback (like the one GAPI gives) using the Promise constructor: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise

Reply:
Yeah you're right. Thank you for the documentation link, I've implemented it with a promise and pushed the fix :).